lib/spack/spack/environment.py

```diff
@@ -557,7 +557,15 @@ def regenerate(self, all_specs, roots):
             if not os.path.isabs(root):
                 root = os.path.normpath(os.path.join(self.base, self.root))
             fs.mkdirp(root)
-            with fs.replace_directory_transaction(root):
+
+            # The tempdir for the directory transaction must be in the same
+            # filesystem mount as the view for symlinks to work. Provide
+            # dirname(root) as the tempdir for the
+            # replace_directory_transaction because it must be on the same
+            # filesystem mount as the view itself. Otherwise it may be
+            # impossible to construct the view in the tempdir even when it can
+            # be constructed in-place.
+            with fs.replace_directory_transaction(root, os.path.dirname(root)):
```

Comment:
I have a couple additional thoughts (sorry this didn't occur to me earlier):

* `os.path.dirname("/x/y/") == "/x/y"`, so we might want to use `os.path.normpath` to ensure that we strip any trailing `/` (that happens conditionally above but is not guaranteed to happen).
* This should work fine for environment views that are created inside the Spack directory (i.e. in `var/spack/environments`) although I'm concerned about environments with customized view locations - we won't always know if the parent directory is the appropriate location to create a temporary directory. I think using the install tree root would be less problematic (the user generally has access to it and I think it would have to be on the same FS as the view or this problem would have already been triggered).
  * Using the install tree root may be problematic for some cases in #11871 but I can think on that more

Reply:
Come to think on it, I think you should always be able to write to the parent directory if you can create the view there in the first place, so my second concern doesn't apply. I think the first one does though - can you normalize the path?